lib/fluent/plugin/filter_kubernetes_metadata.rb

```diff
@@ -274,35 +250,70 @@ def configure(conf)
+    def get_metadata_for_record(match_data, cache_key, create_time)
+      namespace_name = match_data['namespace']
+      pod_name = match_data['pod_name']
       metadata = {
-        'container_name' => container_name,
+        'container_name' => match_data['container_name'],
         'namespace_name' => namespace_name,
-        'pod_name'       => pod_name,
+        'pod_name'       => pod_name
       }
       if @kubernetes_url.present?
-        cache_key = "#{namespace_name}_#{pod_name}"
-
-        this = self
-        pod_metadata = @cache.getset(cache_key) {
-          @stats.bump(:pod_cache_miss)
-          md = this.get_pod_metadata(
-            namespace_name,
-            pod_name,
-          )
-          md
-        }
+        pod_metadata = get_pod_metadata(cache_key, namespace_name, pod_name, create_time)
         metadata.merge!(pod_metadata) if pod_metadata
+      end
+      metadata
+    end
 
-        if @include_namespace_metadata
-          namespace_metadata = @namespace_cache.getset(namespace_name) {
-            @stats.bump(:namespace_cache_miss)
-            get_namespace_metadata(namespace_name)
+    def get_pod_metadata(key, namespace_name, pod_name, record_create_time)
+        ids = @id_cache.getset(key) do
+          @stats.bump(:id_cache_miss)
+          pod_metadata = fetch_pod_metadata(namespace_name, pod_name)
+          namespace_metadata = fetch_namespace_metadata(namespace_name)
+          ids = {:pod_id=> pod_metadata['pod_id'], :namespace_id => namespace_metadata['namespace_id'] }
+          if ids[:pod_id].nil? && !ids[:namespace_id].nil?
+             ns_time = Time.parse(namespace_metadata['creation_timestamp'])
+             ids[:pod_id] = key if ns_time <= record_create_time #ns is older then record
+          end
+          if ids[:pod_id].nil?
+             @stats.bump(:pod_cache_orphaned_record)
+             ids = :orphaned
+          else
+             @cache[ids[:pod_id]] = pod_metadata
+             @namespace_cache[ids[:namespace_id]] = namespace_metadata
+          end
+          ids
+        end
+        if ids == :orphaned
+          return {} unless @allow_orphans
+          log.trace("orphaning message for: #{namespace_name}/#{pod_name} ") if log.trace?
+          return {
+            'orphaned_namespace' => namespace_name,
+            'namespace_name' => @orphaned_namespace_name,
+            'namespace_id' => @orphaned_namespace_id
           }
-          metadata.merge!(namespace_metadata) if namespace_metadata
         end
-      end
-      metadata
+        metadata =  @cache.getset(ids[:pod_id]) do
+            @stats.bump(:pod_cache_miss)
+            m = fetch_pod_metadata(namespace_name, pod_name)
+            (m.nil? || m.empty?) ? nil : m
+        end || {}
+        metadata.merge!(@namespace_cache.getset(ids[:namespace_id]) do
+           @stats.bump(:namespace_cache_miss)
+           m = fetch_namespace_metadata(namespace_name)
+          (m.nil? || m.empty?) ? nil : m
```

Comment:
I don't think we want to cache `nil` if we can't fetch the namespace information.

First, at this point, we already have the fixed container ID to pod ID and namespace ID, so we do know this container's namespace ID.  So we can still return a record with the namespace ID in it.   And for that matter, we can return the namespace name we have, just not in a metadata field that says it "is" the namespace name of that ID.